21599 - Updated Step One Tick Mark Issue

ArwenQin opened the pull request:
*Issue #:* /bcgov/entity#21599

*Description of changes:*
- Step one will only show the Tick Mark and errors, after the user visits the Review and Confirm Page
- The authorization date picker will reset, if the incorporation date is changes to a date after the current selected authorization date. (Otherwise the component will be valid regardless of the invalid authorization date)


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

ArwenQin commented:
I updated the ContinuationInAuthorization component - The authorization date picker will reset, if the incorporation date changes to a date after the current selected authorization date. 
Otherwise the component will be valid regardless of the invalid authorization date, current:
![image](https://github.com/bcgov/business-create-ui/assets/122495122/45709753-bae2-4aba-91fc-6d09a4cf4f5f)

ArwenQin commented:
/gcbrun

bcregistry-sre commented:
Temporary Url for review: https://business-create-dev--pr-705-mufawei3.web.app

severinbeauvais commented:
> I updated the ContinuationInAuthorization component - The authorization date picker will reset, if the incorporation date changes to a date after the current selected authorization date. Otherwise the component will be valid regardless of the invalid authorization date

I don't think the ManualBusinessInfo component should change another component's object.

You can probably make the original code work properly by checking the form validity in ContinuationAuthorization.vue::onComponentValid() (and maybe you need a variable for form validity and you need to watch it, same as [here](https://github.com/bcgov/business-create-ui/blob/fb729e6e879eb7d6ff7163d34a5f51eced1ec402/src/components/ContinuationIn/ManualBusinessInfo.vue#L387).

severinbeauvais commented on src/components/ContinuationIn/ExtraproRegistration.vue:
```diff
@@ -653,13 +654,17 @@ export default class ExtraproRegistration extends Mixins(DateMixin) {
+    // show tick mark only when user visits Review Page
+    if (this.getShowErrors) {
+      return (
+        this.isBusinessActive &&
```
This is a good solution.

Another way to do this would be to move `getShowErrors` inside the return statement.

ArwenQin replied:
Thanks! Updated.

ArwenQin commented:
/gcbrun

ArwenQin commented:
> > I updated the ContinuationInAuthorization component - The authorization date picker will reset, if the incorporation date changes to a date after the current selected authorization date. Otherwise the component will be valid regardless of the invalid authorization date
> 
> ~I don't think the ManualBusinessInfo component should change another component's object.~ _My mistake, I read your comment before I looked at your code. This isn't an issue._
> 
> You can probably make the original code work properly by checking the form validity in ContinuationAuthorization.vue::onComponentValid() (and maybe you need a variable for form validity and you need to watch it, same as [here](https://github.com/bcgov/business-create-ui/blob/fb729e6e879eb7d6ff7163d34a5f51eced1ec402/src/components/ContinuationIn/ManualBusinessInfo.vue#L387)).

Got it, thanks! I updated the coding, much simpler now. Will re-validate the component when the minAuthoDate changes.

bcregistry-sre commented:
Temporary Url for review: https://business-create-dev--pr-705-mufawei3.web.app

severinbeauvais commented on src/components/ContinuationIn/ContinuationAuthorization.vue:
```diff
@@ -292,15 +292,17 @@ export default class ExtraproRegistration extends Mixins(DocumentMixin) {
     // and at least one file uploaded
     return (
       !!this.authorization.date &&
+      new Date(this.minAuthorizationDate) <= new Date(this.authorization.date) &&
```
This is a duplicate of the rules on lines 196-199.

My idea was to use a v-model variable on the v-form. I think onGetShowErrors() will automatically run when minAuthorizationDate changes, which will call the validation functions. Then, here, you can just check if the form is valid.

ArwenQin replied:
> This is a duplicate of the rules on lines 196-199.
> 
> My idea was to use a v-model variable on the v-form. I think onGetShowErrors() will automatically run when minAuthorizationDate changes, which will call the validation functions. Then, here, you can just check if the form is valid.

The form will declare itself as valid even when the authorizationDate is invalid. It seems the formRef.validate() will be valid even when all fields are empty - is it because all sub custom components don't support form validation?

severinbeauvais replied:
Oh, you're on to something.

The DatePicker is one of our custom components, so no, it doesn't support form validation here... but internally it does (since it uses its own form). 

On lines 289 and 290, you call a public method in that component. There is also `isDateValid()` that you might be able to call to check if it's valid.

severinbeauvais replied:
https://github.com/bcgov/bcrs-shared-components/blob/83cfe9c8dbe19714567814eb3efd7e08f35faf49/src/components/date-picker/DatePicker.vue#L127

severinbeauvais replied:
Let me know if this doesn't work, and we'll look for a better or simpler solution.

ArwenQin replied:
Thank you Severin.
I used a const authorizationDateValid to track the authorization date form validity. It works.

ArwenQin commented on src/components/ContinuationIn/ContinuationAuthorization.vue:
```diff
@@ -280,27 +281,30 @@ export default class ExtraproRegistration extends Mixins(DocumentMixin) {
       // validate the form and our custom components that don't support form validation
       this.$refs.formRef.validate()
-      this.$refs.authorizationDateRef.validateForm()
+      this.authorizationDateValid = this.$refs.authorizationDateRef.validateForm()
```
I use this const to track the auth. date validity and watch it in the componentValid method

bcregistry-sre commented:
Temporary Url for review: https://business-create-dev--pr-705-mufawei3.web.app

ArwenQin commented on src/components/ContinuationIn/ContinuationAuthorization.vue:
```diff
@@ -280,27 +281,30 @@ export default class ExtraproRegistration extends Mixins(DocumentMixin) {
   @Watch('getShowErrors')
   @Watch('minAuthorizationDate') // because Authorization Date depends on this
   @Watch('authorization.date') // because Expiry Date depends on this
+  @Watch('getExistingBusinessInfo.homeIncorporationDate') // because Authorization Date validity depends on this
```
In Extrapro situation, if the user select a home incorporation date later than BC registry date:
the min authorization date will be BC registry date.
But the auth. date rule will still check if the authorization date >=  home incorporation date

severinbeauvais replied:
Are you sure you need this? I think that maybe line 282 will trigger this event already.

(Line 282 watches a computed variable, which _should_ auto-update if homeIncorporationDate changes.)

ArwenQin replied:
> Are you sure you need this? I think that maybe line 282 will trigger this event already.
> 
> (Line 282 watches a computed variable, which _should_ auto-update if homeIncorporationDate changes.)

When the min authorization date is set to be bcFoundingDate (line 182). In that case, the change of homeIncorporateDate will not change the min authorization date. 
But the rule (line 199) still checks whether the authorization date >= home incorporation date.

ArwenQin replied:
current screenshot: when homeIncorporateDate changes, the min authorization date does not change
and there will be an error
![image](https://github.com/bcgov/business-create-ui/assets/122495122/c481ccd3-6a9a-4ace-bf57-b37d9f38f3c3)
![image](https://github.com/bcgov/business-create-ui/assets/122495122/3a3d2ea6-3c6c-4b57-9f40-2146378bf3ab)

severinbeauvais replied:
> When the min authorization date is set to be bcFoundingDate (line 182). In that case, the change of homeIncorporateDate will not change the min authorization date. But the rule (line 199) still checks whether the authorization date >= home incorporation date.

I tested this locally. The watcher on line 282 does indeed fire, but the condition on line 286 is false, so the components are not validated. I think this is actually what we want: we do not show errors until the review step has been displayed (and getShowErrors is True).

ArwenQin replied:
If I remove line 284, even when getShowErrors is True - if we change the date of incorporation, the error will not show
![image](https://github.com/bcgov/business-create-ui/assets/122495122/1db26d86-ec94-4546-87b2-38d4bde3ef16)
the error will only show when we try to change the authorization date again:
![image](https://github.com/bcgov/business-create-ui/assets/122495122/e0e5f839-eb0d-4bf0-b793-53b68968077e)

I tested with:
http://localhost:8080/continuation-in-business-home?id=ThevF8a4zF&accountid=3302
I changed the date of incorporation to a date later than authorization date.
It is because when minAuthorizationDate is set to be the bcFoundingDate (line 182). In that case, the change of homeIncorporateDate will not change the min authorization date, so line 282 didn't fire.

severinbeauvais replied:
In your example, the minimum authorization date is 1968-09-10. This comes from the expro record.

Test again but using a manually entered business.

severinbeauvais replied:
You can debug and step through the code to see what the variable values are, or do something like this:

![image](https://github.com/bcgov/business-create-ui/assets/10002889/a88be025-3c06-4ffe-93e1-1ba689d22647)

ArwenQin replied:
> In your example, the minimum authorization date is 1968-09-10. This comes from the expro record.
> 
> Test again but using a manually entered business.

Yes, the minimum authorization date (1968-09-10) is the bcFoundingDate (line 182), and irrelevant to the homeIncorporateDate. 
So if we only watch the @Watch('minAuthorizationDate') (line 282), when the user changes the homeIncorporateDate (say to June 11 2024), the minAuthorizationDate won't change, so line 285 - 294 didn't run. The current authorization date is actually invalid (Jun 3 2024), but the error will not show, and the tick mark incorrectly shows.

severinbeauvais replied:
OK, Arwen and I sorted this out. (Needed to add another rule.) Looks good to me now. This conversation can be resolved.

severinbeauvais commented on src/components/ContinuationIn/ContinuationAuthorization.vue:
```diff
@@ -280,27 +281,30 @@ export default class ExtraproRegistration extends Mixins(DocumentMixin) {
     // and at least one file uploaded
     return (
       !!this.authorization.date &&
+      this.authorizationDateValid &&
```
You don't need line 306 :)

ArwenQin commented:
/gcbrun

ArwenQin commented:
/gcbrun

bcregistry-sre commented:
Temporary Url for review: https://business-create-dev--pr-705-mufawei3.web.app

bcregistry-sre commented:
Temporary Url for review: https://business-create-dev--pr-705-mufawei3.web.app

ArwenQin commented:
/gcbrun

bcregistry-sre commented:
Temporary Url for review: https://business-create-dev--pr-705-mufawei3.web.app

severinbeauvais commented:
@ArwenQin , thanks for all your testing and trying things, and pushing back when I was missing something key. Together, we got this! :)
